torch/_higher_order_ops/utils.py

```diff
@@ -0,0 +1,21 @@
+from typing import Any
+
+import torch
+import torch.utils._pytree as pytree
+
+
+def autograd_not_implemented_check(args: Any, op_name: str) -> None:
+    """If autograd is enabled and any of the operands require grad this will raise an error with message.
+
+    Args:
+        operands (List[Any]): The flattened operands to the HigherOrderOperator
+        message (str): The name of the HigherOrderOperator
+
+    Raises:
+        RuntimeError: If autograd is enabled and any of the operands require grad
+    """
+    flat_operands, _ = pytree.tree_flatten(args)
+    if torch.is_grad_enabled() and any(
+        f.requires_grad for f in flat_operands if isinstance(f, torch.Tensor)
+    ):
+        raise RuntimeError(f"Autograd is not supported for {op_name}")
```

Comment:
We want this to be a delayed error like how PyTorch built-in operators will get a delayed error (autogradNotImplementedFallback) codegen'ed. Example over at https://github.com/pytorch/pytorch/blob/646fa36875821f2bcf4fbfbf669c1f4f9f69700d/functorch/experimental/_cond.py#L138-L140

The delayed error will raise an error in the backward pass. It's possible that an operator is used with inputs that require grad but .backward() never actually gets called, we want to allow those situations.

Reply:
(and we should also use this mechanism to replace the autograd implementations for cond, the rng HigherOrderOps (https://github.com/pytorch/pytorch/blob/646fa36875821f2bcf4fbfbf669c1f4f9f69700d/torch/_prims/rng_prims.py#L171)